torch/csrc/jit/codegen/cuda/lower_insert_syncs.cpp

```diff
@@ -68,244 +72,255 @@ class SmemAllocMap {
+        if (!alloc_stack.back().sync_after_read &&
+            !alloc_stack.back().sync_before_write) {
+          insert_sync = within_iter_loop_;
+        }
 
-  //! Shared Memory Reads at end of the for-loop
-  //! Cleared after each SyncThreads
-  TvSet final_;
+        alloc_stack.pop_back();
+        if (alloc_stack.empty()) {
+          to_erase.push_back(entry.first);
+        }
+      }
+    }
 
-  //! Track first sync deterministically found in for-loop. Even when a
-  //! child loop has a sync, if it may not be executed due to non-zero
-  //! start value, this flag remains false.
-  bool initial_sync_ = false;
+    for (auto tv : to_erase) {
+      smem_allocations_.erase(tv);
+    }
 
-  //! Track if last op is sync
-  bool is_last_op_sync_ = false;
-};
+    // WAR Sync is necessary in this loop, register its insertion.
+    if (insert_sync) {
+      kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+      auto sync_expr = ir_builder.create<kir::Sync>(true);
+      kir::ExprMutator::registerInsertAfter(
+          for_loop->body().exprs().back(), sync_expr, &for_loop->body());
+      handle(sync_expr);
+    }
 
-class LocalSyncInserter {
- public:
-  //! Write-After-Read race conditions are only found within for-loops.
-  //! Sync nodes are inserted directly into the for-loops.
-  //! The expressions are modified in-place and exprs is const.
-  static void insertSyncs(const std::vector<kir::Expr*>& exprs) {
-    LocalSyncInserter inserter;
-    inserter.insert(exprs);
+    // Pop for loop scope information
+    sync_hit_.pop_back();
+    within_iter_loop_ = prev_within_iter_loop_;
   }
 
- private:
-  void insert(const std::vector<kir::Expr*>& exprs) {
-    for (auto expr : exprs) {
-      if (auto fl = dynamic_cast<kir::ForLoop*>(expr)) {
-        LocalSyncInserterForLoop sync_inserter(fl, alloc_map_);
-      } else if (auto ite = dynamic_cast<kir::IfThenElse*>(expr)) {
-        insert(ite->thenBody().exprs());
-        insert(ite->elseBody().exprs());
-      } else if (auto alloc = dynamic_cast<kir::Allocate*>(expr)) {
-        alloc_map_.insert(alloc);
-      }
+  // Create a new WarMemoryInfo entry if required and return a reference to it,
+  // else return the WarMemoryInfo associated with tv
+  WarMemoryInfo& getMemInfo(kir::TensorView* tv) {
+    auto maybe_aliased_tv = alloc_map_.getRealBuffer(tv);
+    auto alloc_it = smem_allocations_.find(maybe_aliased_tv);
+    auto ca_loop = loop_utils::getAllocInformation(tv->fuserTv(), for_loops_)
+                       .init_for_loop;
+    if (alloc_it == smem_allocations_.end()) {
+      WarMemoryInfo mem_info;
+      mem_info.ca_loop = ca_loop;
+      auto entry_it =
+          smem_allocations_
+              .insert(std::make_pair(
+                  maybe_aliased_tv, std::vector<WarMemoryInfo>({mem_info})))
+              .first;
+      return entry_it->second.back();
+    } else if (
+        maybe_aliased_tv != tv && alloc_it->second.back().ca_loop != ca_loop) {
```

Comment:
When does this hit? No C++ test seems to hit.

Reply:
None of the Python tests seems to hit either.

Reply:
I'm not certain it's needed. I was worried the aliasing pass could alias memory with different compute at loops, which is at least theoretically possible but maybe isn't supported. This might be more likely if we had store_at. I think I'd prefer leaving it in for now, but soft tagging @shmsong in case he remembers the alias pass better than I.